File: android/src/main/kotlin/com/cleveradssolutions/plugin/flutter/bridge/base/EventHandler.kt
package com.cleveradssolutions.plugin.flutter.bridge.base

import androidx.annotation.CallSuper
import androidx.annotation.MainThread
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import io.flutter.plugin.common.EventChannel

abstract class EventHandler(
    private val channelName: String
) : EventChannel.StreamHandler {

    private var channel: EventChannel? = null
    private var eventSink: EventChannel.EventSink? = null

    @CallSuper
    @MainThread
    open fun onAttachedToFlutter(flutterPluginBinding: FlutterPluginBinding) {
        channel = EventChannel(flutterPluginBinding.binaryMessenger, channelName).also {
            it.setStreamHandler(this)
        }
    }

    @CallSuper
    @MainThread
    open fun onDetachedFromFlutter() {
        channel?.let {
            it.setStreamHandler(null)
            channel = null
        }
    }

    @CallSuper
    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    @CallSuper
    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    protected fun success(vararg pairs: Pair<*, *>) {
        eventSink?.success(pairs.toMap())
    }

}

File: ios/Classes/Bridge/Base/EventHandler.swift
//
//  EventHandler.swift
//  clever_ads_solutions
//
//  Copyright © 2024 CleverAdsSolutions LTD, CAS.AI. All rights reserved.
//

import Flutter

class EventHandler: NSObject, FlutterStreamHandler {
    private var channelName: String
    private var channel: FlutterEventChannel?
    private var eventSink: FlutterEventSink?

    init(channelName: String) {
        self.channelName = channelName
    }

    open func onAttachedToFlutter(_ registrar: FlutterPluginRegistrar) {
        let binaryMessenger = registrar.messenger()
        channel = FlutterEventChannel(name: channelName, binaryMessenger: binaryMessenger)
        channel!.setStreamHandler(self)
    }

    open func onDetachedFromFlutter() {
        channel?.setStreamHandler(nil)
        channel = nil
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }

    internal func success(_ map: [String: Any?]) {
        eventSink?(map)
    }
}
